ui/src/app/shared/components/dropdown/dropdown.tsx

```diff
@@ -19,112 +20,107 @@ export interface DropDownState {
 
 require('./dropdown.scss');
 
-const dropDownOpened = new BehaviorSubject<DropDown>(null);
-
-export class DropDown extends React.Component<DropDownProps, DropDownState> {
-    private el: HTMLDivElement;
-    private content: HTMLDivElement;
-    private subscriptions: Subscription[];
-
-    constructor(props: DropDownProps) {
-        super(props);
-        this.state = {opened: false, left: 0, top: 0};
-    }
-
-    public render() {
-        let children: React.ReactNode = null;
-        if (typeof this.props.children === 'function') {
-            if (this.state.opened) {
-                const fun = this.props.children as () => React.ReactNode;
-                children = fun();
-            }
-        } else {
-            children = this.props.children as React.ReactNode;
-        }
+const dropDownOpened = new BehaviorSubject<typeof DropDown>(null);
 
-        return (
-            <div className='argo-dropdown' ref={el => (this.el = el)}>
-                <div
-                    qe-id={this.props.qeId}
-                    className='argo-dropdown__anchor'
-                    onClick={event => {
-                        this.open();
-                        event.stopPropagation();
-                    }}>
-                    <this.props.anchor />
-                </div>
-                {ReactDOM.createPortal(
-                    <div
-                        className={classNames('argo-dropdown__content', {'opened': this.state.opened, 'is-menu': this.props.isMenu})}
-                        style={{top: this.state.top, left: this.state.left}}
-                        ref={el => (this.content = el)}>
-                        {children}
-                    </div>,
-                    document.body
-                )}
-            </div>
-        );
-    }
+export function DropDown({isMenu, anchor: Anchnor, children, qeId}: DropDownProps) {
+    const anchorRef = useRef<null | HTMLDivElement>(null);
+    const contentRef = useRef<null | HTMLDivElement>(null);
+    const subscriptionsRef = useRef<null | Subscription[]>(null);
+    const [opened, setOpened] = useState(false);
+    const [left, setLeft] = useState(0);
+    const [top, setTop] = useState(0);
 
-    public componentWillMount() {
-        this.subscriptions = [
+    useEffect(() => {
+        const content = contentRef.current;
+        const el = anchorRef.current;
+
+        subscriptionsRef.current = [
             merge(
-                dropDownOpened.pipe(filter(dropdown => dropdown !== this)),
+                dropDownOpened.pipe(filter(dropdown => dropdown !== DropDown)),
```

Comment:
hmm this function and `.next` might be the problem. did using `this` not work? functions still have `this`

I think it might make sense to just remove the `filter` entirely by moving `dropDownOpened` into the component itself as a ref. The `filter` is only needed because there could be multiple instantiations of this component and `dropDownOpened` is currently at module-level. If we move `dropDownOpened`  into component-level, then each instantiation will have their own ref -- i.e. `dropDownOpened` becomes component-local.

I think then the generic can be removed as well and it could just accept no arguments, i.e. `dropDownOpened.next()`.

Also Rx is not really needed for these simple event listeners, so we could even go a step further and remove Rx from this component entirely.

Making `dropDownOpened` a component-level ref is the easier change, so I'd probably suggest you do that unless you want to rewrite more 😅

Reply:
Yes this doesn't work I tested and found that the filter function this in the class component returns the value of the DropDown object
<img width="663" alt="image" src="https://github.com/argoproj/argo-workflows/assets/100748721/92006d5b-44f6-4130-9030-5e435adc579c">
but in the functional component it returns a function and this returns undefined. 🤔
<img width="651" alt="image" src="https://github.com/argoproj/argo-workflows/assets/100748721/e615b850-3d6f-4d04-94c4-492e90e351f6">
If I could pass this function as an object to `dropDownOpened`, I could save the existing code, but I can't think of a way to do it easily 😅 I moved `dropDownOpened` inside the component as you said, but `close()` still doesn't work 🥲 For now, I'm trying to figure out if there's a way to pass this function component as an object to `dropDownOpened.pipe`!

Reply:
> but in the functional component it returns a function and this returns undefined. 🤔

Yea the former is what I expected. You were passing the function itself, whereas the class was passing an _instantiation_.

I thought the latter might work since all functions are also objects in JS, but I appear to have forgotten some of the [detailed nuances there](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/this). That is a hacky method though, removing the filter entirely with a component-local ref would be better anyway.

Looks like you managed to figure out how to remove Rx entirely though in your recent commit (nice work!), so seems like this is moot now?

Reply:
Yes when I tested it, both close and open work fine! Even when I clicked inside the filter list, there is no event propagation and everything looks fine 👍 I'll re-commit with the details you mentioned, thanks for your help as always 🙇‍♂️